Hide bookmarks link for the blocks in the LTI mode

dmitry-viskov opened the pull request:
There is no need to show bookmarks link for the blocks that are displayed through the LTI.
Because there is no way for users accessing content to reach the bookmarks list.

![canvas1](https://cloud.githubusercontent.com/assets/1876893/19654176/fd73ca0e-9a1f-11e6-9701-bc86437cebe4.jpeg)

![canvas2](https://cloud.githubusercontent.com/assets/1876893/19654184/014cf8bc-9a20-11e6-9feb-393cabaaf06a.jpeg)

edx-webhook commented:
Thanks for the pull request, @dmitry-viskov! I've created [OSPR-1527](https://openedx.atlassian.net/browse/OSPR-1527) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

marcotuts commented:
It makes sense to me that from a product standpoint the bookmarks action shouldn't be rendered in a different (LTI) context. Moving to engineering review next.

robrap commented:
Hi @dmitry-viskov.  Could you add a test for the hidden bookmark?  Please add a comment to the test that the LTI view, which uses the "chromeless" view, should not support bookmarking.  Thanks.

dmitry-viskov commented:
jenkins run python

dmitry-viskov commented:
jenkins run js

dmitry-viskov commented:
@robrap done

robrap commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -206,3 +207,23 @@ def test_unauthenticated(self):
+            )
+        self.setup_user(admin=False, enroll=False, login=True)
+        response = self.get_response(usage_id=unicode(self.vetical_block.location))
+        self.assertNotContains(response, '<div class="bookmark-button-wrapper"')
```
Thanks for adding the test.  

I'd like the test to be a bit more resilient to changes in the implementation of bookmarks.  One possibility is to use a variable like `bookmark_html = '<div class="bookmark-button-wrapper"'` in include a case where bookmarks do appear (i.e. `self.assertContains(response, bookmark_html)`).  Thus, if implementation changed, the test would fail.  I know that LTI might be an awkward place for this opposite condition though.

Another possibility which is far less fool proof, but makes it slightly more resilient, would be to just search for "bookmark".

Thoughts?

dmitry-viskov replied:
ok. I've added more more detailed check

robrap commented:
Thanks @dmitry-viskov.  Also, it is nicer for reviewers if you wait and squash at the very end.  That makes it easier for me to know that there are no other changes that came in while you were back in the code.

dmitry-viskov commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -206,3 +207,23 @@ def test_unauthenticated(self):
+        """
+        self.setup_course()
+        chapter = ItemFactory.create(parent=self.course, category='chapter')
+        self.vetical_block = ItemFactory.create(  # pylint: disable=attribute-defined-outside-init
```
Fixed

dmitry-viskov commented:
@robrap I've changed code according your comments and made rebase. Please check

dmitry-viskov commented:
@robrap any news according this PR?

robrap commented:
Sorry for the delay @dmitry-viskov.

Again, please do not amend or squash your commits during the review process.  Add a new commit to fix any review feedback so it is simpler to see what has changed.  Thank you.

For the test, I'm now thinking a couple of things:

- Maybe the test would be better in edx-platform/common/lib/xmodule/xmodule/tests/test_vertical.py?
- The test can then just focus on `test_render_chromeless_view`.  It would need to use `assert_bookmark_info_not_in`.
- You could keep a comment about LTI in the test, although you may want to note the chromeless view is also used for Mobile.
- You probably need a constant to make `assert_bookmark_info_in` and `assert_bookmark_info_not_in` stay in sync.
- Hopefully the chromeless test code setup can be a little simpler, but not sure.
- You should also be testing for the actual test HTML, as is done in the test_verical.py test.

Let me know if this all makes sense.  Thanks.

dmitry-viskov commented:
@robrap thanks for the answer.

>Again, please do not amend or squash your commits during the review process. 

Other your colleagues usually asks me to make rebase each time ;-)

>Maybe the test would be better in edx-platform/common/lib/xmodule/xmodule/tests/test_vertical.py?

To be honest for me it sounds like overkill development. The described problem is solved and not directly linked with vertical blocks. In my opinion it does not matter where to place this test. It cover this bug and possible use cases. Don't think that anything here should be changed especially given the fact that i've spent time to create this test.

>You probably need a constant to make assert_bookmark_info_in and assert_bookmark_info_not_in stay in sync.

I think for this simple case specially dedicated functions like `assert_bookmark_info_in` and `assert_bookmark_info_not_in` are not needed

>You should also be testing for the actual test HTML, as is done in the test_verical.py test.

Don't understand what do you mean. We need to solve particular problem. Not cover all other separate cases

robrap commented:
Thanks for your comments @dmitry-viskov .  I'll address each individually.

> Other your colleagues usually asks me to make rebase each time ;-)

Sorry you are getting different directions from different developers.  You can see in our [TNL PR template](https://openedx.atlassian.net/wiki/display/TNL/TNL+Pull+Request+Template) that we have "Rebase and squash commits" as a post-review task.  I'm not sure if the previous request you got was for squashing before the PR has begun, or another scenario, or in direct conflict with what I am saying. :)  If you point me to a PR or specific colleague, we could clear it up so you get more consistent instructions or you and I can better understand why there is a conflict.

 > To be honest for me it sounds like overkill development.

I'm sorry that my request would cause some refactoring, but let me explain more of my thinking:

- As written, the current test is really difficult to understand what is being tested and why.  This is especially related to the new setup code.
- I then looked to the existing test code to see if some cleaner setup code could be used.  And what I noticed was:
  - The existing bookmark test was not only simpler to follow, but it already exists (as would be expected).
  - It would be better to avoid redundant code/tests, since tests are very difficult to maintain when we consider the entire platform, and not a single test.
  - My original request was around keeping close proximity between the with and without bookmarks tests, which I now think would be better served by moving the new test near the existing test.
  - Although the requirement is related to LTI (thus the need for the LTI-specific comment), the implementation is all inside vertical_block with no mention of LTI.  That is ok, but if I were a future developer looking to adjust that code in vertical_block, my expectation would be that its test coverage would be in the test file for the same, which is exactly where the bookmark test appears.

For these reasons, I think that moving/shrinking the test would be the best and simplest solution to cleaning up the current test.  The next best solution, but one I don't particularly like, would be to go with a more generic assertion for "bookmark", assuming that there isn't bookmark related HTML even in the non-bookmark view.  I definitely prefer `test_vertical.py` at this time though.  It may be very simple to do, since there is already test data that just needs to be displayed in the chromeless view. 

> I think for this simple case specially dedicated functions like assert_bookmark_info_in and assert_bookmark_info_not_in are not needed.

This comment of mine assumed that the test would be moved where `assert_bookmark_info_in` already existed.  For the "chromeless-view" test, it could use `assert_bookmark_info_not_in` to be more symmetric, or it could potentially inline the assertion.  There are several assertions inside `assert_bookmark_info_in`, and I would hope that `not_in` would be simpler.  It may even just be able to check that "bookmark" doesn't appear.  It would be great if the two assertions were somehow related, noting that your assertion for `bookmark-button-wrapper` doesn't currently appear in `assert_bookmark_info_in`.

> Robert: You should also be testing for the actual test HTML, as is done in the test_verical.py test.
> Dmitry: Don't understand what do you mean. We need to solve particular problem. Not cover all other separate cases

I was referring to the equivalent [to this assertion in `test_vertical.py`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/tests/test_vertical.py#L75).  Your test adds `"<p>Test HTML Content %d<p>"` to the content.  If the new test asserts that I do see this test HTML, and I don't see the bookmarks, it would be a simple improvement to the test to ensure that the bookmark assertion is run against a valid response.  Also, if the test were moved and you could use the existing test data, then it would literally be adding this exact line to the new test.

I know I wrote a lot here.  If you continue to disagree with my proposal, which you may, maybe we could find a quick time to chat. 

Thanks again.

dmitry-viskov commented:
@robrap 

I disagree with you. `test_vertical.py` is used for testing low level render of some fragments. All `common/lib/xmodule` code designed not for the business logic. It is some kind of internal libraries that are used in the particular places in the LMS and CMS. If you print for example [this](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/tests/test_vertical.py#L72) `html` variable you will see something like that:
```
(('vert_module.html',
  {'bookmark_id': u'bilbo,i4x://edX/xml_test_course/vertical/2',
   'bookmarked': False,
   'items': [{'content': u'Test HTML 1',
              'id': u'i4x://edX/xml_test_course/html/test-html-1'},
             {'content': u'Test HTML 2',
              'id': u'i4x://edX/xml_test_course/html/test-html-2'}],
   'show_bookmark_button': True,
   'unit_title': '2',
   'xblock_context': {'bookmarked': False,
                      'show_bookmark_button': False,
                      'username': 'bilbo'}}),
 {})
```
The type of this variable is `string`. So this test is designed for testing render vertical blocks in case of base `STUDENT_VIEW`, `AUTHOR_VIEW`, `STUDIO_VIEW` and etc.

What you call `chromeless view` is some kind of high level django view: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/views.py#L1228

We can't test in the low level test case logic from some particular django view. Of course I may create test like this:

```
    def assert_bookmark_info_not_in(self, content):
        """
        Assert content has no bookmark info.
        """
        self.assertIn('bookmark_id', content)
        self.assertIn('{},{}'.format(self.username, unicode(self.vertical.location)), content)
        self.assertIn("'show_bookmark_button': False", content)

    def test_render_chromeless_view(self):

        self.module_system._services['bookmarks'] = Mock()  # pylint: disable=protected-access
        self.module_system._services['user'] = StubUserService()  # pylint: disable=protected-access

        context = copy(self.default_context)
        context['show_bookmark_button'] = False
        html = self.module_system.render(self.vertical, STUDENT_VIEW, context).content

        self.assertIn(self.test_html_1, html)
        self.assertIn(self.test_html_2, html)
        self.assert_bookmark_info_not_in(html)
```
But it is wrong and ugly of course. Because it is not the test of chromeless view. It is representation of `STUDENT_VIEW` with `show_bookmark_button == False` param passed in the context.

dmitry-viskov commented:
@robrap I suggest not to invent new bicycle. The test that I've written already cover case with display in the chromeless view. If you want I may also add test to cover direct call of the `courseware.views.views.render_xblock` view. For example in the new class that inherits this one: https://github.com/edx/edx-platform/blob/b2cd0359e5d11e0c0dae50c69c0f477ab7a40932/lms/djangoapps/courseware/tests/test_views.py#L1999

But if you still think that `test_vertical.py` is the best place for this - please provide me your own solution how do you want this test was written

robrap commented:
@dmitry-viskov: Sorry if I was mistaken about `test_vertical.py`.  I'm not at my computer and we have a 2-day hackathon starting tomorrow.

Do you know the proper test file for both the file you are changing that is testing at the proper level for chromeless view?  That is the file that might be the proper location, but I couldn't say without reviewing what is there.

I'm concerned with 3 things:
- clarity of the test
- correctness and resilience of the test 
- location of the test

Based on what you wrote, I was pointing to a worse location. There is a reasonable argument for where the test currently lies, and I understand that. I was curious if the chromeless view test class had any related code that would make this easier to find and avoid some redundancy, and that may not be the case.

If after reviewing this other test file we decide it is in the best location under LTI Provider, I think that just checking that the chromeless view does not contain 'bookmark' (assuming this is accurate, because I don't know if the bookmarkless view is 100% free of the term) would make it more resilient. This is one of the options I first gave.

The inclusion of the non-chromeless code just makes the existing test to difficult to decipher. I know I proposed it and I thank you for implementing it. This is just my feedback on the implentation. I know it could also be cleaned up with helpers and/or comments, but I don't think that is worth it.

I would still add the assertion for the test HTML as I mentioned.

Thanks for disagreeing when you disagree. ;) That is much better than blindly implementing the wrong solution. 

Let me know what you think is the best way to resolve clean up based on what I wrote and what you're current thinking is.

dmitry-viskov commented:
@robrap I've added `TestRenderVerticalXBlock` test with check "contains" and modified `TestRenderXBlock.verify_response` to check "not contains". It is much better solution.
The existed `LtiLaunchTestRender.test_hidden_bookmarks` should also stay.

dmitry-viskov commented:
jenkins run bokchoy

robrap commented:
@dmitry-viskov: It turns out that the right location for this test is [COURSEWARE_CHROME_HTML_ELEMENTS  in RenderXBlockTestMixin](https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/lms/djangoapps/courseware/testutils.py#L30) is the right location for HTML that should only appear when in the Chrome view.

I would use this, and remove all other tests.  I would add a comment like `# Bookmarks must not be appear in LTI`.  I do not think the test code should be duplicated under LTI Provider.

Thank you.

dmitry-viskov commented:
@robrap I disagree with you. `COURSEWARE_CHROME_HTML_ELEMENTS` is only for `HTML/Problem` blocks. `RenderXBlockTestMixin` is designed to verify `HTML` elements. But `bookmarks` refer to 'Vertical` blocks

dmitry-viskov commented:
@robrap there is one test `TestRenderVerticalXBlock.test_bookmark_available` to check rendering Vertical XBlock with bookmarks, one test `TestRenderXBlock.test_vertical_block_chromeless_view` to check that vertical block in chromeless_view doesn't contain bookmarks and I still think that LTI test also should stay.

I do not see any reason to continue improve this PR.

robrap commented:
@dmitry-viskov: I'd like the tests as clean as possible is because there is a large cost to maintaining all of them.  I continue to advocate for using `COURSEWARE_CHROME_HTML_ELEMENTS` in RenderXBlockTestMixin, but I now see that the setup will need to also have a vertical, and the block under test would need to be switched to the vertical.  As with my other recommendations, this is an untested theory.  I have the day off, and may try this myself when I return.  The idea is that this would replace all other test code changes.

I didn't want to leave you with no response for another day.  I'll take this up on Thursday from wherever you leave it.

Thanks for your efforts.

robrap commented:
@dmitry-viskov: When you get a chance, let me know what your plan is so I know whether or not I need to attempt to implement my recommendation.  Thanks!

dmitry-viskov commented:
@robrap 

>When you get a chance, let me know what your plan is so I know whether or not I need to attempt to implement my recommendation. Thanks!

Hi. It would be nice if you provide your solution

dmitry-viskov commented:
@robrap big thanks for your assistance!

robrap commented:
@dmitry-viskov You're welcome.
